### Pandas

In this notebook we are going to learn some `pandas` stuff when working with data. In this notebook we are going to look at the following in pandas.

1. Input and Output ✔
2. Pandas series ✔
3. DataFrames ✔
4. Pandas Arrays ✔
5. Index Objects ✔


If you don't have pandas installed you can install it by running the following command:

```shell
!pip install pandas
```

> Note that when you are using `anacoda` you don't need to install pandas as it comes installed.

In the following code cell we are going to import `pandas` with an alias `pd` and check the version of pandas.

In [1]:
import pandas as pd
pd.__version__

'1.5.3'

### Input and Output 

In this section we are going to look at `io` in pandas. Which is basically reading and writing in pandas.

1. `CSV` files - these are files columns of data are serperated using a comma, that why it's called `Comma Serperated Values`. We are going to read and write to csv files.

> To read a `csv` file in pandas we use the `.read_csv` function. This function returns us a pandas dataframe.

In [2]:
dataframe = pd.read_csv('files/airquality.csv')

In the `dataframe` object we can basically check the first `5` rows of data using the `head()` function.

In [3]:
dataframe.head()

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0           1   41.0    190.0   7.4    67      5    1
1           2   36.0    118.0   8.0    72      5    2
2           3   12.0    149.0  12.6    74      5    3
3           4   18.0    313.0  11.5    62      5    4
4           5    NaN      NaN  14.3    56      5    5

In the `dataframe` object we can basically check the last `5` rows of data using the `tail()` function.

In [4]:
dataframe.tail()

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
148         149   30.0    193.0   6.9    70      9   26
149         150    NaN    145.0  13.2    77      9   27
150         151   14.0    191.0  14.3    75      9   28
151         152   18.0    131.0   8.0    76      9   29
152         153   20.0    223.0  11.5    68      9   30

> In the `head()` or `tail()` function you can specify the number of rows you want.

In [5]:
dataframe.head(10)

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0           1   41.0    190.0   7.4    67      5    1
1           2   36.0    118.0   8.0    72      5    2
2           3   12.0    149.0  12.6    74      5    3
3           4   18.0    313.0  11.5    62      5    4
4           5    NaN      NaN  14.3    56      5    5
5           6   28.0      NaN  14.9    66      5    6
6           7   23.0    299.0   8.6    65      5    7
7           8   19.0     99.0  13.8    59      5    8
8           9    8.0     19.0  20.1    61      5    9
9          10    NaN    194.0   8.6    69      5   10

In [6]:
dataframe.tail(7)

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
146         147    7.0     49.0  10.3    69      9   24
147         148   14.0     20.0  16.6    63      9   25
148         149   30.0    193.0   6.9    70      9   26
149         150    NaN    145.0  13.2    77      9   27
150         151   14.0    191.0  14.3    75      9   28
151         152   18.0    131.0   8.0    76      9   29
152         153   20.0    223.0  11.5    68      9   30

> Note that all the data that we are working with comes inform of `csv` files. We might also have files like:

1. excel spreadsheet
2. text files (.txt)
3. json files
4. sql files
5. etc.


The `read_excel` method is used to read an excel spreadsheep into a pandas dataframe. In our files directory we have an excel file that is called `cars.xlsx` and we want to read this file in a pandas dataframe. We can do it as follows.




In [7]:
df = pd.read_excel('files/cars.xlsx')
df.head()

mpg  engine  horse  weight  accel  year  origin  cylinder  filter_.   mpg1
0   9.0     4.0   93.0     732    8.5     0     NaN       NaN       NaN   9.98
1  10.0   360.0  215.0    4615   14.0    70     1.0       8.0       0.0  10.87
2  10.0   307.0  200.0    4376   15.0    70     1.0       8.0       0.0   9.63
3  11.0   318.0  210.0    4382   13.5    70     1.0       8.0       0.0  12.12
4  11.0   429.0  208.0    4633   11.0    72     1.0       8.0       0.0  10.63

> You can use some dataframe methods like `head()`, `tail()` etc on the `df` object because we know that `read_excel` returns us a pandas dataframe.

Pandas exposed a method called `read_json()` which allows us to read `json` data in `json` file into a pandas dataframe. We have a `json` file that is located in the `files` directory and it's name is `todos.json`. Let's read this file to a pandas dataframe.

In [8]:
df = pd.read_json('files/todos.json')
df.head(6)

userId  id                                              title  completed
0       1   1                                 delectus aut autem      False
1       1   2                 quis ut nam facilis et officia qui      False
2       1   3                                fugiat veniam minus      False
3       1   4                                   et porro tempora       True
4       1   5  laboriosam mollitia et enim quasi adipisci qui...      False
5       1   6  qui ullam ratione quibusdam voluptatem quia omnis      False

In pandas we can read `txt` files using the `read_csv`. When reading text files we need to specify the delimeter for our columns in the `read_csv` method. We have a file called `files/airquality.txt` that contains the data that looks as follows:

```txt
Ozone   Solar.R Wind    Temp    Month   Day
41  190 7.4 67  5   1
36  118 8   72  5   2
12  149 12.6    74  5    3
18  313 11.5    62  5    4
```

We will read this file into a pandas dataframe using the `read_csv` and specify the delimeter to be a `\t` because the columns are sepurated by tabs.

In [9]:
df = pd.read_csv('files/airquality.txt', delimiter='\t')
df.head()

Ozone   Solar.R Wind    Temp    Month   Day
0                       41  190 7.4 67  5   1
1                       36  118 8   72  5   2
2                  12  149 12.6    74  5    3
3                  18  313 11.5    62  5    4

### Dataframe

In this section we are going to have a look at how we can create our own dataframe from iteratable using the `DataFrame` class. Let's start by creating a dataframe from `list` of `dictionaries` as follows:

In [10]:
users = [
    {'gender': 'male', 'name': 'Jonh', 'age': 56, 'surname': 'Doe'},
    {'gender': 'female', 'name': 'Mary', 'age': 26, 'surname': 'Jack'},
    {'gender': 'male', 'name': 'Peter', 'age': 34, 'surname': 'Gross'}
]

df = pd.DataFrame(users)
df

gender   name  age surname
0    male   Jonh   56     Doe
1  female   Mary   26    Jack
2    male  Peter   34   Gross

We can also use `list` of `lists` to create a pandas dataframe. Let's have a look at the following example. 

> When using `list` of `lists`  we need to specify the `columns` which are the column names.

In [11]:
rows = [user.values() for user in users]
headers = users[0].keys()
headers, rows

(dict_keys(['gender', 'name', 'age', 'surname']),
 [dict_values(['male', 'Jonh', 56, 'Doe']),
  dict_values(['female', 'Mary', 26, 'Jack']),
  dict_values(['male', 'Peter', 34, 'Gross'])])

In [12]:
df = pd.DataFrame(rows, columns=headers)
df

gender   name  age surname
0    male   Jonh   56     Doe
1  female   Mary   26    Jack
2    male  Peter   34   Gross

The dataframe object has various a methods of saving data to files. Let's save our dataframes to files in a folder called `saved`.

In [13]:
import os

if not os.path.exists('saved'):
    os.mkdir('saved')

Let's save our dataframe to a `.csv` file in teh folder called `saved`.

> Specifying that `index=False` we are telling pandas that we don't want a column of index in our `csv` file.

In [14]:
df.to_csv('saved/people.csv', index=False)
print('Saved')

Saved


We can use the `to_excel` to convert a dataframe to an excel spreadsheet for exmaple.

In [15]:
df.to_excel('saved/people.xlsx', index=False)
print('Saved')

Saved


We can use the `to_json` to convert a dataframe to a `json` file for exmaple. We need to specify the `orient`.

If we want a our json file to have a list of dictionaries we need to set the `lines` to be false else the output in the file will be lines of dictionaries.

In [24]:
df.to_json('saved/people.json', orient='records', lines=False, indent=2)
print('Saved')

Saved


### DataFrame object

In this section we are going to look at some methods and properties that we can use on the `DataFrame` class. We have looked at the `head()` and the `tail()` methods which checks for the first  and last `5` rows or data in the dataframe.

> You can pass `n` to the  `head()` and the `tail()` methods to specify number of columns you want in a dataframe.


We have a method called `reset_index()`. This method is responsible for resteing the index in a dataframe object.

In [25]:
dataframe.tail(3).reset_index(drop=True)

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0         151   14.0    191.0  14.3    75      9   28
1         152   18.0    131.0   8.0    76      9   29
2         153   20.0    223.0  11.5    68      9   30

We can access the columns of a dataframe by calling the `columns` property.

In [26]:
columns = dataframe.columns
columns

Index(['Unnamed: 0', 'Ozone', 'Solar.R', 'Wind', 'Temp', 'Month', 'Day'], dtype='object')

We can get the values using the `values` property.

In [80]:
dataframe.values

array([[ 41.        , 190.        ,   7.4       , ...,   5.        ,
          1.        ,  19.44444444],
       [ 36.        , 118.        ,   8.        , ...,   5.        ,
          2.        ,  22.22222222],
       [ 12.        , 149.        ,  12.6       , ...,   5.        ,
          3.        ,  23.33333333],
       ...,
       [ 14.        , 191.        ,  14.3       , ...,   9.        ,
         28.        ,  23.88888889],
       [ 18.        , 131.        ,   8.        , ...,   9.        ,
         29.        ,  24.44444444],
       [ 20.        , 223.        ,  11.5       , ...,   9.        ,
         30.        ,  20.        ]])

In pandas dataframe we can access the values in the `column` that we are intrested in using the `values` propety. But first we need to index the column name that we want data from first. 

> Calling values will return a numpy version of values. Calling `dataframe['Month']` or `dataframe.Month` returns us a pandas [Series](https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series), inorder for us to get values in form of numpy we need to call the `.values` for example:

In [32]:
months = dataframe['Month'].values
months

array([5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
       5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
       6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,
       8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9, 9, 9,
       9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9],
      dtype=int64)

We can call the `unique()` method on a pandas series to get the unique values of that column.

In [33]:
months = dataframe['Month'].unique()
months

array([5, 6, 7, 8, 9], dtype=int64)

Let's say we want to get the first `5` temperatures or last `5` temperatures in our dataframe. We can do it as follows:

In [35]:
dataframe['Temp'].head()

0    67
1    72
2    74
3    62
4    56
Name: Temp, dtype: int64

In [37]:
dataframe['Temp'].tail(2).reset_index(drop=True)

0    76
1    68
Name: Temp, dtype: int64

In our dataframe we have a column called `Unnamed: 0` Let's go ahead and drop or remove that column from our dataframe. We can use the `drop` method. In the drop method we can specify a kwarg called `inplace` by default it is set to false. Setting it to true will modify the dataframe inplace.

In [44]:
dataframe.drop(columns=['Unnamed: 0'], inplace=True)

In [45]:
dataframe.head()

Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5

We can select columns that we want in a dataframe using `indexing`. Let's demostrate by sellecting `Month` and `Temp` columns from our dataframe.

In [59]:
dataframe[['Month', 'Temp']].head()

Month  Temp
0      5    67
1      5    72
2      5    74
3      5    62
4      5    56

If i want to get the rows i can use the `loc` or `iloc` method to index rows.

In [60]:
dataframe.iloc[1]

Ozone       36.0
Solar.R    118.0
Wind         8.0
Temp        72.0
Month        5.0
Day          2.0
Name: 1, dtype: float64

In [61]:
dataframe.loc[1]

Ozone       36.0
Solar.R    118.0
Wind         8.0
Temp        72.0
Month        5.0
Day          2.0
Name: 1, dtype: float64

In [63]:
dataframe.iloc[:3]

Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3

We can create a column based on another column in a dataframe. We want to create a column called `Degrees`. In this column we want to take the `Temp` column and convert it from `F` to `C`.

In [66]:
dataframe['Degrees'] = (dataframe['Temp']-32)*5/9
dataframe.head()

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees
0   41.0    190.0   7.4    67      5    1  19.444444
1   36.0    118.0   8.0    72      5    2  22.222222
2   12.0    149.0  12.6    74      5    3  23.333333
3   18.0    313.0  11.5    62      5    4  16.666667
4    NaN      NaN  14.3    56      5    5  13.333333

We have a attribute called `at` which allows us to index the value at a specified `row` and at a specified `column` name. Let's access the value `23.333333` from our dataframe. We can do it as follows:

In [68]:
dataframe.at[2, 'Degrees']

23.333333333333332

We can use the `iat` attribute to acess the value using tyhe row index and column index.

In [78]:
dataframe.iat[2, 6]

23.333333333333332

We can check the datatypes for each column in a dataframe using the `dtypes` as follows:

In [70]:
dataframe.dtypes

Ozone      float64
Solar.R    float64
Wind       float64
Temp         int64
Month        int64
Day          int64
Degrees    float64
dtype: object

We can also check the shape of the dataframe using the `shape` attribute as follows:

In [71]:
dataframe.shape # 153 rows and 7 columns

(153, 7)

We can check if the dataframe is `empty` using the `empty` attribute as follows.

In [74]:
empty_df = pd.DataFrame()

In [76]:
dataframe.empty, empty_df.empty

(False, True)

We can check the how many dimention does the dataframe have using the `ndim` attribute.

In [79]:
dataframe.ndim

2

We can check the size of the dataframe using the `size` property as follows.

In [81]:
dataframe.size

1071

### Refs

1. [pandas.pydata.org](https://pandas.pydata.org/docs/user_guide/index.html)
2. []()